src/components/application_manager/src/event_engine/event_dispatcher_impl.cc

```diff
@@ -125,36 +116,32 @@ void EventDispatcherImpl::remove_observer(EventObserver& observer) {
   }
 }
 
-void EventDispatcherImpl::remove_observer_from_vector(EventObserver& observer) {
-  observers_.erase(
-      std::remove_if(
-          observers_.begin(), observers_.end(), IdCheckFunctor(observer.id())),
-      observers_.end());
-}
-
 // Mobile Events
 
 void EventDispatcherImpl::raise_mobile_event(const MobileEvent& event) {
-  AutoLock observer_lock(mobile_observer_lock_);
+  ObserverVector mobile_observers;
 
-  // check if event is notification
-  if (mobile_apis::messageType::notification == event.smart_object_type()) {
-    const uint32_t notification_correlation_id = 0;
-    mobile_observers_ =
-        mobile_observers_event_[event.id()][notification_correlation_id];
-  }
+  {
+    AutoLock observer_lock(mobile_observer_lock_);
```

Comment:
On second thought, the latter portion of this method is modifying `mobile_observers_event_` via this code:
```
while (!mobile_observers.empty()) {
    EventObserver* temp = *mobile_observers.begin();
    mobile_observers.erase(mobile_observers.begin());
```

Should `observer_lock` be held until after the `while (!mobile_observers.empty())` loop?
If so, the same should be applied to `raise_event` as well.